File: python-algorithm/algo_05_dynamic_plan/dp_3_two_sequence/L583-m.cpp
//
// Created by LiHeng on 2020/6/21.
//

//这道题转化成求最长公共子序列问题，参考1143
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size() ;
        int n = word2.size() ;
        int longest = 0 ;
        vector<vector<int>> memo(m+1 ,vector<int> (n+1,0));
        //memo[0][0]存储的是2个字符串中都没有字符的状态
        //memo[i][j]代表word1[0:i) word2[0:j) 两个字符串的最长公共子序列
        for(int i=1 ; i < m+1 ; i++){
            for(int j = 1 ; j < n+1 ;j++){
                if(word1[i-1] == word2[j-1]){
                    memo[i][j] = memo[i-1][j-1] + 1 ;
                }
                else
                    memo[i][j] = max(memo[i-1][j], memo[i][j-1]) ;
            }
        }
        return m+n-2*memo[m][n];
    }
};

File: python-algorithm/algo_05_dynamic_plan/dp_3_two_sequence/L583-m.py
#--coding:utf-8--


# 这道题仔细分析发现和1143求最长公共子序列其实是一回事，操作数等于字母总数-2*长度
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        dp = [[0]*(n+1) for i in range(m+1)]
        for i in range(1, m+1):
            for j in range(1, n+1):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]+1
                else:
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])
        return m+n-2*dp[m][n]


# 这道题也可以用编辑距离72的思路，状态转移方程要相应变一变，因为没了替换操作。
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        dp = [[0]*(n+1) for i in range(m+1)]
        for i in range(m+1):
            dp[i][0] = i
        for j in range(n+1):
            dp[0][j] = j
        for i in range(1, m+1):
            for j in range(1, n+1):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    # +2 表示两个字符串各删除一次的操作
                    dp[i][j] = min(min(dp[i-1][j], dp[i][j-1])+1, dp[i-1][j-1]+2)
        return dp[m][n]